File: mpp/lab8/client_java/src/main/java/rest/RacesResource.java
package rest;

import domain.Arbiter;
import domain.Race;

import javax.ws.rs.client.Entity;
import javax.ws.rs.client.WebTarget;
import javax.ws.rs.core.GenericType;
import javax.ws.rs.core.MediaType;
import java.util.List;

public class RacesResource {
    private final API api;

    public RacesResource(API api) {
        this.api = api;
    }

    public Race add(String name, Integer arbiterId) {
        WebTarget target = api.getTarget().path("/races");
        Race race = new Race(name, new Arbiter(arbiterId));
        return target
                .request(MediaType.APPLICATION_JSON_TYPE)
                .post(Entity.json(race), Race.class);
    }

    public Race findById(Integer id) {
        WebTarget target = api.getTarget().path("/races/" + id);
        return target
                .request(MediaType.APPLICATION_JSON_TYPE)
                .get(Race.class);
    }

    public List<Race> find() {
        WebTarget target = api.getTarget().path("/races");
        return target
                .request(MediaType.APPLICATION_JSON_TYPE)
                .get(new GenericType<>(){});
    }

    public Race update(Integer id, String name, Integer arbiterId) {
        WebTarget target = api.getTarget().path("/races/" + id);
        Race race = new Race(name, new Arbiter(arbiterId));
        return target
                .request(MediaType.APPLICATION_JSON_TYPE)
                .post(Entity.json(race), Race.class);
    }

    public Race delete(Integer id) {
        WebTarget target = api.getTarget().path("/races/" + id);
        return target
                .request(MediaType.APPLICATION_JSON_TYPE)
                .delete(Race.class);
    }
}

File: mpp/lab8/client_web/web/js/RacesResource.js
import axios from 'axios';

export default class {
    constructor(api) {
        this.api = api;
    }

    getPath(sub_path) {
        let path = this.api.getPath('/races');
        if (sub_path) {
            path += sub_path;
        }
        return path;
    }

    getData(race) {
        const data = {};
        if (race.name) {
            data.name = race.name;
        }
        if (race.arbiter && race.arbiter.id) {
            data.arbiter = {};
            data.arbiter.id = race.arbiter.id;
        }
        return data;
    }

    async add(race) {
        const response = await axios.post(this.getPath('/'), this.getData(race));
        return response.data;
    }

    async find() {
        const response = await axios.get(this.getPath('/'));
        return response.data;
    }

    async update(race) {
        const response = await axios.patch(this.getPath(`/${race.id}`), this.getData(race));
        Object.assign(race, response.data);
        return response.data;
    }

    async delete(id) {
        await axios.delete(this.getPath(`/${id}`));
    }
}
